stdlib/private/OSLog/OSLogMessage.swift

```diff
@@ -0,0 +1,309 @@
+      bitWidth: bitWidth,
+      isSigned: isSigned)
+
+    addArgumentHeaders(
+      // Always explicitly set either public or private flag bit.
+      flagAndType: isPrivate ? 1 : 2,
+      byteSize: UInt8(bitWidth / 8))
+
+    updateSummaryBytes(isPrivate: isPrivate)
+  }
+
+  // Update the summary bytes: preamble and argument count.
+  public mutating func updateSummaryBytes(isPrivate: Bool) {
+    preamble |= isPrivate ? 1 : 0
+    argumentCount += 1
+  }
+
+  // Append the given argument header and size.
+  public mutating func addArgumentHeaders(flagAndType: UInt8, byteSize: UInt8) {
+    arguments.append(flagAndType)
+    arguments.append(byteSize)
+  }
+
+  /// Construct an OS log format specifier from the given parameters.
+  /// All arguments to this function must be known at compile time.
+  public func getIntegerFormatSpecifier(
+    _ format: IntFormat,
+    isPrivate: Bool,
+    bitWidth: Int,
+    isSigned: Bool
+  ) -> String {
+    var formatSpecifier: String = isPrivate ? "%{private}" : "%{public}"
+
+    // Determine the length modifier of the format specifier.
+    switch bitWidth {
+    case _ where bitWidth == CInt.bitWidth:
+      break
+    case _ where bitWidth == CLongLong.bitWidth:
+      formatSpecifier += "ll"
+    default:
+      fatalError("Int size not supported")
+    }
+
+    switch (format) {
+    case .hex:
+      formatSpecifier += "x"
+    case .octal:
+      formatSpecifier += "o"
+    default:
+      formatSpecifier += isSigned ? "d" : "u"
+    }
+    return formatSpecifier
+  }
+}
+
+extension String {
+  /// Replace all percents "%" in the string by "%%" so that the string can be
+  /// interpreted as a C format string.
+  public var percentEscapedString: String {
+    get {
+      return self
+        .split(separator: "%", omittingEmptySubsequences: false)
+        .joined(separator: "%%")
+    }
+  }
+}
+
+public struct OSLogMessage :
+  ExpressibleByStringInterpolation, ExpressibleByStringLiteral
+{
+  public let oslogInterpolation : OSLogInterpolation
+
+  /// Initializer for accepting string interpolations.
+  public init(stringInterpolation: OSLogInterpolation) {
+    oslogInterpolation = stringInterpolation
+  }
+
+  /// Initializer for accepting string literals.
+  public init(stringLiteral value: String) {
+    // Note that the actual value of `literalCapacity` is not important as it
+    // is ignored by `OSLogInterpolation.init`. However, it must be a literal.
+    var s = OSLogInterpolation(literalCapacity: 1, interpolationCount: 0)
+    s.appendLiteral(value)
+    self.init(stringInterpolation: s)
+  }
+
+  /// Format string constructed from the string interpolation.
+  public var formatString: String {
+    get { return oslogInterpolation.formatString }
+  }
+
+  /// The byte size of the argument buffer that will contain the elements of
+  /// `oslogInterpolation.arguments` and the two summary bytes: preamble and
+  /// argument count.
+  public var argumentBufferByteSize: Int {
```

Comment:
Should this just be "byteCount" or "bufferSize" or "messageSize" or "messageSizeInBytes"? Since this includes the bytes of the summary, calling it the "argumentBuffer" seems like it could be misleading.